isa.c

```diff
@@ -293,7 +298,89 @@ static void exec_B(uint32_t instruction)
         next_state.regs[LR] = curr_state.regs[PC] + 4;
         // address of instruction after the branch instruction stored in LR
     }
-    next_state.regs[PC] = curr_state.regs[PC] + (sign_extend(signed_immed_24,30,24) << 2);
+    next_state.regs[PC] = curr_state.regs[PC] + (int32_t)(sign_extend(signed_immed_24, 24, 30) << 2);
 }
 
+static void exec_BL(uint32_t instruction)
+{
+    if (condition_check(curr_state, get_bits(instruction, 31, 28))) {
+        uint32_t immed_24 = get_bits(instruction, 23, 0);
+        uint8_t L = get_bit(instruction, 24);
+        if (L == 1) {
+            next_state.regs[LR] = curr_state.regs[PC] + 4;
+        }
+        next_state.regs[PC] += (int32_t)(sign_extend(immed_24, 24, 30) << 2);
+    }
+}
+
+static void exec_RSB(uint32_t instruction)
+{
+    struct ShifterOperand * shiftop;
+    shiftop = shifter_operand(curr_state, instruction);
+    uint8_t Rdi = get_bits(instruction, 15, 12);
+    uint8_t Rni = get_bits(instruction, 19, 16);
+    uint8_t S = get_bit(instruction, 20);
+    if (condition_check(curr_state, get_bits(instruction, 31, 28))) {
+        next_state.regs[Rdi] = shiftop->shifter_operand - curr_state.regs[Rni];
+        if (S == 1) { //ignore the SPSR crap.
+            set_bit(&next_state.CPSR, CPSR_N, get_bit(next_state.regs[Rdi], 31));
+            set_bit(&next_state.CPSR, CPSR_Z, !next_state.regs[Rdi]);
+            set_bit(&(curr_state.CPSR), CPSR_C, !check_sub_borrow(shiftop->shifter_operand, curr_state.regs[Rni]));
+            set_bit(&(curr_state.CPSR), CPSR_V, !check_overflow(shiftop->shifter_operand, -curr_state.regs[Rni]));
```

Comment:
yah change this everywhere, along with the curr_state/next_state errors in all four lines